src/main/java/org/reactome/release/orthopairs/UniProtGeneNamesRetriever.java

```diff
@@ -0,0 +1,178 @@
+            if (partition.size() == MAX_UNIPROT_BATCH_QUERY_SIZE) {
+                partitionedUniProtIds.add(partition);
+                partition = new HashSet<>();
+            }
+        }
+        // The final identifiers at the end are added to the List here.
+        partitionedUniProtIds.add(partition);
+        return partitionedUniProtIds;
+    }
+
+    /**
+     * Query UniProt via the Java API for gene names using UniProt accession IDs.
+     * @param partitionedUniProtIds List<Set<String>> Partitioned UniProt identifiers.
+     * @return Set<String> - Accession-to-Name mappings that will be stored to a file.
+     * @throws ServiceException - Thrown by UniProtService class if service is unavailable.
+     * @throws InterruptedException - Thrown if the Sleep process that occurs after every batch query is interrupted.
+     */
+    public static Set<String> retrieveGeneNamesFromUniProt(List<Set<String>> partitionedUniProtIds) throws ServiceException, InterruptedException {
+
+        // Create the UniProtService.
+        ServiceFactory serviceFactoryInstance = Client.getServiceFactoryInstance();
+        UniProtService uniprotService = serviceFactoryInstance.getUniProtQueryService();
+        uniprotService.start();
+        int count = 0;
+        Set<String> uniprotAccessionsToGeneNames = new HashSet<>();
+        for (Set<String> uniprotIdentifierPartition : partitionedUniProtIds) {
+            // Build UniProt API query from Set of 250 UniProt identifiers.
+            Query query = UniProtQueryBuilder.accessions(uniprotIdentifierPartition);
+            // Perform UniProt API query to retrieve gene names associated with identifiers.
+            QueryResult<UniProtComponent<Gene>> uniprotEntries = uniprotService.getGenes(query);
+
+            while (uniprotEntries.hasNext()) {
+                count++;
+                // Get Gene object returned from UniProt.
+                UniProtComponent<Gene> geneObject = uniprotEntries.next();
+                if (!geneObject.getComponent().isEmpty()) {
+                    // Iterate through all Gene components in the response.
+                    for (Gene geneComponent : geneObject.getComponent()) {
+                        // Tab-separate UniProt accession ID and its associated gene name, and then store these in the Set that will be returned.
+                        uniprotAccessionsToGeneNames.add(geneObject.getAccession().toString() + "\t" + geneComponent.getGeneName().toString() + "\n");
+                    }
+                }
+
+                if (count % 1000 == 0) {
+                    logger.info(count + " UniProt identifiers have been processed");
+                }
+            }
+            // UniProt has a 200 requests/second limit for their API. Chances are we don't need this restriction, but its polite!
+            final long QUERY_SLEEP_DURATION = Duration.ofSeconds(2).toMillis();
```

Comment:
Have you measured the number of _actual_ requests/second the application sends?

Reply:
It might be possible to speed the process up by reducing this sleep, or having the application be aware of when it gets close to 200 requests / second.